crates/ruff_python_parser/tests/snapshots/invalid_syntax@statements__function_type_parameters.py.snap

```diff
@@ -372,17 +372,40 @@ Module(
                     body: [
                         Expr(
                             StmtExpr {
-                                range: 1036..1039,
-                                value: EllipsisLiteral(
-                                    ExprEllipsisLiteral {
-                                        range: 1036..1039,
+                                range: 1028..1031,
+                                value: NumberLiteral(
+                                    ExprNumberLiteral {
+                                        range: 1028..1031,
+                                        value: Int(
+                                            100,
+                                        ),
                                     },
                                 ),
                             },
                         ),
                     ],
                 },
             ),
+            AnnAssign(
+                StmtAnnAssign {
+                    range: 1032..1039,
+                    target: Tuple(
+                        ExprTuple {
+                            range: 1032..1034,
+                            elts: [],
+                            ctx: Store,
+                            parenthesized: true,
+                        },
+                    ),
+                    annotation: EllipsisLiteral(
+                        ExprEllipsisLiteral {
+                            range: 1036..1039,
+                        },
+                    ),
+                    value: None,
+                    simple: false,
+                },
+            ),
```

Comment:
This is not an ideal recovery but it's happening because we removed `skip_until`. Here's an explanation of why's that the case (from Discord):

---

Removing `skip_while` has some consequences. I was trying to figure out why the type parameter snapshot is changing for the following code:

```py
def foo[A,,100](): ...
```

In the original PR, the parser would recover correctly and provide a proper AST containing the function node which didn't have 100 as a type parameter but now the `(): ...` is parsed as an annotated assignment statement.

So, our recovery context when the parser is trying to parse a list of type parameters:
- Module statements
  - Type parameters

Now, when the parser sees `100`, which is _not_ the start of a type parameter (not in FIRST set), it checks if it can be recovered via an outer context. The outer context being module statement, it can because `100` is a start of a statement, specifically a simple statement. This means that the parser is at `100` in module statement context but as there's no semicolon, it'll create an error (as expected).

The problem comes when we're trying to parse the function parameters. Earlier there was a use of `skip_while` if a token was found which is not a parameter then we'd skip until the end of parameter list. But now as we don't do it, we move ahead being in the context of module statement and parse `]` as error because we expect a statement and then `(): ...` as annotated assign statement.

This is correct but I might look into statement parsing to see if it can be improved.